contracts/plugins/trading/DutchTradeRouter.sol

```diff
@@ -0,0 +1,125 @@
+    using SafeERC20 for IERC20;
+    struct Bid {
+        /// @notice The DutchTrade that was bid on
+        DutchTrade trade;
+        /// @notice The token sold to the protocol
+        IERC20 sellToken;
+        /// @notice The amount of tokenIn the protocol got
+        uint256 sellAmt;
+        /// @notice The token bought from the trade
+        IERC20 buyToken;
+        /// @notice The amount of tokenOut the we got
+        uint256 buyAmt;
+    }
+
+    /// @notice Emitted when a bid is placed
+    /// @param main The main contract of the rToken
+    /// @param trade The DutchTrade that was bid on
+    /// @param bidder The address of the bidder
+    /// @param sellToken The token being sold by the protocol
+    /// @param soldAmt The amount of sellToken sold
+    /// @param buyToken The token being bought by the protocol
+    /// @param boughtAmt The amount of buyToken bought
+    event BidPlaced(
+        IMain main,
+        DutchTrade trade,
+        address bidder,
+        IERC20 sellToken,
+        uint256 soldAmt,
+        IERC20 buyToken,
+        uint256 boughtAmt
+    );
+    DutchTrade private _currentTrade = DutchTrade(address(0));
+
+    /// Place a bid on an OPEN dutch auction
+    /// @param trade The DutchTrade to bid on
+    /// @param recipient The recipient of the tokens out
+    /// @dev Requires msg.sender has sufficient approval on the tokenIn with router
+    /// @dev Requires msg.sender has sufficient balance on the tokenIn
+    function bid(DutchTrade trade, address recipient) external returns (Bid memory) {
+        Bid memory out = Bid({
+            trade: DutchTrade(address(0)),
+            sellToken: IERC20(address(0)),
+            sellAmt: 0,
+            buyToken: IERC20(address(0)),
+            buyAmt: 0
+        });
+        _placeBid(trade, out, msg.sender);
+        _sendBalanceTo(out.sellToken, recipient);
+        _sendBalanceTo(out.buyToken, recipient);
+        return out;
+    }
+
+    /// @notice Callback for DutchTrade
+    /// @param caller The caller of the callback, should be the router
+    /// @param buyToken The token DutchTrade is expecting to receive
+    /// @param buyAmount The amt the DutchTrade is expecting to receive
+    /// @notice Data is not used here
+    function dutchTradeCallback(
+        address caller,
+        address buyToken,
+        uint256 buyAmount,
+        bytes calldata
+    ) external {
+        require(caller == address(this), "Invalid caller");
+        require(msg.sender == address(_currentTrade), "Incorrect callee");
+        IERC20(buyToken).safeTransfer(msg.sender, buyAmount);
+    }
+
+    function _sendBalanceTo(IERC20 token, address to) internal {
+        uint256 bal = token.balanceOf(address(this));
+        if (bal == 0) {
+            return;
+        }
+        token.safeTransfer(to, bal);
+    }
+
+    // Places a bid on a Dutch auction
+    // Method will dynamically pull funds from msg.sender if needed
```

Comment:
I don't think we want to use "if needed" logic here given that the router will be used by many different parties. It should always transfer in exactly the number of tokens needed, imo, else we're just asking for a low-information party to send tokens to the contract before getting frontrun by an MEV bot during the actual bid.

Reply:
Yeah I'll remove that branch. It was a leftover from when I envisioned supporting placing multiple bids at the same time via the UI.